File: Power Query/300-399/313/PQ 313.py
import pandas as pd
import numpy as np

path = "300-399/313/PQ_Challenge_313.xlsx"
input = pd.read_excel(path, usecols="A:B", nrows=18)
test = pd.read_excel(path, usecols="D:E", nrows=4)

pivoted = input.assign(group=(input['Data1'] == "Customer").cumsum()) \
    .pivot(index='group', columns='Data1', values='Data2') \
    .reset_index(drop=True)
pivoted.iloc[:, 1:] = pivoted.iloc[:, 1:].apply(pd.to_numeric, errors='coerce')

pivoted['Total Amount'] = (
    pivoted['Quantity'] * (
        pivoted['Price']
        - pivoted.filter(like='Disc').sum(axis=1, skipna=True)
        - pivoted.filter(like='Tax').sum(axis=1, skipna=True)
    )
).astype('int64')

result = pivoted[['Customer', 'Total Amount']]
result.loc[len(result)] = ['Total Sale', result['Total Amount'].sum()]

print(test.equals(result)) # True

File: Power Query/300-399/313/PQ 313.R
library(tidyverse)
library(readxl)
library(janitor)

path = "Power Query/300-399/313/PQ_Challenge_313.xlsx"
input = read_excel(path, range = "A1:B18")
test  = read_excel(path, range = "D1:E5")

result = input %>%
  mutate(group = cumsum(Data1 == "Customer")) %>%
  pivot_wider(
    names_from = Data1,
    values_from = Data2,
    values_fill = NA
  ) %>%
  mutate(across(-c(group, Customer), as.numeric)) %>%
  mutate(`Total Amount` = Quantity * (Price -rowSums(across(starts_with("Disc")), na.rm = T) - rowSums(across(starts_with("Tax")), na.rm = T))) %>%
  select(Customer, `Total Amount`) %>%
  adorn_totals("row", name = "Total Sale")
  
all.equal(result, test, check.attributes = FALSE) 
# > TRUE  